wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+        return false;
+    }
+
+    return setupWiFi();
+}
+
+bool ESPNOWBroadcast::setupWiFi() {
+    Serial.println("ESPNOWBroadcast::setupWiFi()");
+
+    _state.exchange(STOPPED);
+
+    // To enable ESPNow, we need to be in WIFI_STA mode
+    if ( !WiFi.mode(WIFI_STA) ) {
+        Serial.println("WiFi.mode() failed");
+        return false;
+    }
+    // and not have the WiFi connect
+    // Calling discount with tigger an async Wifi Event
+    if ( !WiFi.disconnect(false, true) ) {
+        Serial.println("WiFi.disconnect() failed");
+        return false;
+    }
+
+    return true;
+}
+
+
+void ESPNOWBroadcast::loop(size_t maxMessagesToProcess /*= 1*/) {
+    switch (espnowBroadcast._state.load()) {
+        case ESPNOWBroadcast::STARTING:
+            // if WiFI is in starting state, actually stat ESPNow from our main task thread.
+            start();
+            break;
+        case ESPNOWBroadcast::STARTED: {
+            auto ndx = maxMessagesToProcess;
+            while(ndx-- > 0) {
+                size_t size = 0;
+                auto *msg = (QueuedNetworkMessage*)xRingbufferReceive(_networkRxMessages, &size, 0);
+                if (msg) {
+                    if(size == sizeof(*msg)) {
+                        auto callback = _rxCallback;
+                        while( *callback ) {
+                            (*callback)(msg->mac, msg->data, msg->len);
+                            callback++;
+                        }
+                    }
+                    vRingbufferReturnItem(_networkRxMessages, (void *)msg);
+                } else {
+                    break;
+                }
+            }
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+ESPNOWBroadcast::err_t ESPNOWBroadcast::send(const uint8_t* msg, size_t len) {
+    static const uint8_t broadcast[] = BROADCAST_ADDR_ARRAY_INITIALIZER;
+    return static_cast<ESPNOWBroadcast::err_t>(esp_now_send(broadcast, msg, len));
+}
+
+void ESPNOWBroadcast::start() {
+
+    Serial.println("starting ESPNow");
+
+    if ( WiFi.mode(WIFI_STA) ) {
+        if ( WiFi.status() == WL_DISCONNECTED ) {
+            if (esp_wifi_start() == ESP_OK) {
+                if (esp_now_init() == ESP_OK) {
+                    if (esp_now_register_recv_cb(onESPNowRxCallback) == ESP_OK) {
+                        static esp_now_peer_info_t peer = {
+                            BROADCAST_ADDR_ARRAY_INITIALIZER,
+                            {0},
+                            WLED_ESPNOW_WIFI_CHANNEL,
+                            WIFI_IF_STA,
+                            false,
+                            NULL
+                            };
+
+                        if (esp_now_add_peer(&peer) == ESP_OK) {
+                            ESPNOWBroadcast::STATE starting {ESPNOWBroadcast::STARTING};
+                            if (espnowBroadcast._state.compare_exchange_strong(starting, ESPNOWBroadcast::STARTED)) {
+                                return;

```

Comment:
I'm curious - atomic variables, just a practice you use generally 
or is there another thread you're worried is going to use the state?

Reply:
I think I answered my own question - `onWifiEvent`?

Reply:
onWiFiEvent happens on a different thread / task.   The default arduino implementation uses underly FreeRTOS queues.  But that is heavyweight for what we actually need